src/leap/mail/imap/account.py

```diff
@@ -248,66 +231,92 @@ def select(self, name, readwrite=1):
         """
         Deletes a mailbox.
 
-        Right now it does not purge the messages, but just removes the mailbox
-        name from the mailboxes list!!!
-
         :param name: the mailbox to be deleted
         :type name: str
 
-        :param force: if True, it will not check for noselect flag or inferior
-                      names. use with care.
+        :param force:
+            if True, it will not check for noselect flag or inferior
+            names. use with care.
         :type force: bool
+        :rtype: Deferred
         """
-        name = self._parse_mailbox_name(name)
+        name = normalize_mailbox(name)
+        _mboxes = None
+
+        def check_it_exists(mailboxes):
+            global _mboxes
+            _mboxes = mailboxes
+            if name not in mailboxes:
+                err = imap4.MailboxException("No such mailbox: %r" % name)
+                return defer.fail(err)
+
+        def get_mailbox(_):
+            return self.getMailbox(name)
 
-        if name not in self.mailboxes:
-            raise imap4.MailboxException("No such mailbox: %r" % name)
-        mbox = self.getMailbox(name)
+        def destroy_mailbox(mbox):
+            return mbox.destroy()
 
-        if force is False:
+        def check_can_be_deleted(mbox):
+            global _mboxes
             # See if this box is flagged \Noselect
-            # XXX use mbox.flags instead?
             mbox_flags = mbox.getFlags()
-            if self.NOSELECT_FLAG in mbox_flags:
+            if MessageFlags.NOSELECT_FLAG in mbox_flags:
                 # Check for hierarchically inferior mailboxes with this one
                 # as part of their root.
-                for others in self.mailboxes:
+                for others in _mboxes:
                     if others != name and others.startswith(name):
-                        raise imap4.MailboxException, (
+                        err = imap4.MailboxException(
                             "Hierarchically inferior mailboxes "
                             "exist and \\Noselect is set")
-        self.__mailboxes.discard(name)
-        mbox.destroy()
-
-        # XXX FIXME --- not honoring the inferior names...
-
+                        return defer.fail(err)
+            return mbox
+
+        d = self.account.list_all_mailbox_names()
+        d.addCallback(check_it_exists)
+        d.addCallback(get_mailbox)
+        if not force:
+            d.addCallback(check_can_be_deleted)
+        d.addCallback(destroy_mailbox)
+        return d
+
+        # FIXME --- not honoring the inferior names...
         # if there are no hierarchically inferior names, we will
         # delete it from our ken.
+        # XXX is this right?
         # if self._inferiorNames(name) > 1:
-        #  ??! -- can this be rite?
-        # self._index.removeMailbox(name)
+        #   self._index.removeMailbox(name)
 
+    # TODO use mail.Account.rename_mailbox
+    # TODO finish conversion to deferreds
```

Comment:
I don't understand this TODOs. If I understand correctly the code is already using mail.Account.rename_mailbox and I don't see what is missing to convert to deferreds.